src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,977 @@
+        /// </summary>
+        protected override QueryExpressionSyntax FindNodeToRefactor(SyntaxNode root, TextSpan span)
+        {
+            var node = root.FindNode(span);
+            return node as QueryExpressionSyntax ?? (node is ArgumentSyntax argument ? argument.Expression as QueryExpressionSyntax : default);
+        }
+
+        private sealed class Converter
+        {
+            private readonly SemanticModel _semanticModel;
+            private readonly ISemanticFactsService _semanticFacts;
+            private readonly CancellationToken _cancellationToken;
+            private readonly QueryExpressionSyntax _source;
+            private readonly List<string> _introducedLocalNames;
+
+            public Converter(SemanticModel semanticModel, ISemanticFactsService semanticFacts, QueryExpressionSyntax source, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _semanticFacts = semanticFacts;
+                _source = source;
+                _introducedLocalNames = new List<string>();
+                _cancellationToken = cancellationToken;
+            }
+
+            public bool TryConvert(out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // Do not try refactoring queries with comments or conditional compilation in them.
+                // We can consider supporting queries with comments in the future.
+                if (_source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    _source.ContainsDirectives))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // Bail out if there is no chance to convert it even with a local function.
+                if (!CanTryConvertToLocalFunction() ||
+                    !TryCreateStackFromQueryExpression(out QueryExpressionProcessingInfo queryExpressionProcessingInfo))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // GetDiagnostics is expensive. Move it to the end if there were no bail outs from the algorithm.
+                // TODO likely adding more semantic checks will perform checks we expect from GetDiagnostics
+                // We may consider removing GetDiagnostics.
+                // https://github.com/dotnet/roslyn/issues/25639
+                if ((TryConvertInternal(queryExpressionProcessingInfo, out documentUpdateInfo) ||
+                    TryReplaceWithLocalFunction(queryExpressionProcessingInfo, out documentUpdateInfo)) && // second attempt: at least to a local function
```

Comment:
>queryExpressionProcessingInfo [](start = 48, length = 29)

Is it possible that TryConvertInternal call above pops something from this stack before failing?

Reply:
> Is it possible that TryConvertInternal call above pops something from this stack before failing?

This comment was marked as resolved, but I didn't see any changes preventing that. Does it mean that the answer to the question is the stack won't be corrupted?

---
In reply to: [178422729](https://github.com/dotnet/roslyn/pull/25362#discussion_r178422729) [](ancestors = 178422729)

Reply:
Sorry! I misread the comment. 

It should not pop anything from the stack before failing. Therefore, we can retry with TryReplaceWithLocalFunction

---
In reply to: [178432034](https://github.com/dotnet/roslyn/pull/25362#discussion_r178432034) [](ancestors = 178432034,178422729)